[FIX+REF] account_move_line_tax_editable: fixes not being able to def…

eantones opened the pull request:
…ine the tax when creating a new journal entry

Before this fix, when a new journal entry was created, the taxes fields were readonly. They were only editable after saving the entry for the first time and editing it again.

Additionally, there's some refactoring too.

eantones commented:
Any feedback about this PR?
Thanks

eantones commented:
Hi @benwillig and @pedrobaeza  any feedback about this PR?
Thank you very much

sbidoul commented:
Why do you remove `_compute_is_tax_editable`? It is important to keep it for situations where people want to override it (eg testing for a user group to allow specific users only to edit taxes).

eantones committed:
[FIX+REF] account_move_line_tax_editable: fixes not being able to define the tax when creating a new journal entry

    Before this fix, when a new journal entry was created, the taxes fields were readonly. They were only editable after saving the entry for the first time and editing it again.

    Additionally, there's some refactoring too.

eantones commented:
Hi @sbidoul ,

ok, I'll keep _compute_is_tax_editable method but as it stands now, it does not always work when creating new records. It is not triggered when creating a new line from account.view_move_form, but it is triggered from account.view_move_line_form. To solve this, it is necessary to force the trigger every time a new account.move.line is created from any form, we can do it with a related field move_state.

I solve this in the current PR, please check it out and tell me what you think.

Just FYI, I test this on the latest commit 10253a735f8296fbb42c311ebc36e2fe4bf1f135

Thanks you very much

sbidoul commented on account_move_line_tax_editable/models/account_move_line.py:
```diff
@@ -6,19 +6,19 @@
     def _get_is_tax_editable(self):
         self.ensure_one()
-        return self.move_id.state == 'draft'
+        return self.move_state in (False, 'draft')
```
`return self.move_id.state == 'draft' or not self.move_id.state` is a tad more robust as it works with different falsy values.

eantones replied:
You're right, but I think ```return not self.move_id or self.move_id.state == 'draft'``` is even better. Thanks

sbidoul replied:
Ah, but then we don't need move_state anymore, right?

eantones replied:
Yes and no, ```move_state``` is there only to trigger ```_compute_is_tax_editable```, without that related field, this method would never triggered when new record was created. 
....But, I thinking it twice.... I think we can achieve the same without a related field :-)

sbidoul commented on account_move_line_tax_editable/views/account_move.xml:
```diff
@@ -1,23 +1,19 @@
         <field name="inherit_id" ref="account.view_move_form"/>
         <field name="arch" type="xml">
             <xpath expr="//field[@name='line_ids']/tree/field[@name='name']" position="after">
+                <field name="move_state" invisible="1"/>
```
Is this line necessary?

eantones replied:
Yes, without this line the computed method is not triggered

sbidoul commented on account_move_line_tax_editable/views/account_move_line.xml:
```diff
@@ -1,16 +1,14 @@
         <field name="inherit_id" ref="account.view_move_line_form"/>
         <field name="arch" type="xml">
             <field name="tax_line_id" position="before">
+                <field name="move_state" invisible="1"/>
```
Is this line necessary?

eantones replied:
Yes, the same as before, without this line the computed method is not triggered

sbidoul replied:
Ah, ok, thanks.

eantones committed:
[IMP] account_move_line_tax_editable: more robust check

eantones committed:
[IMP] account_move_line_tax_editable: remove move_state

sbidoul commented:
Thanks!

eantones commented:
Thank you!

pedrobaeza commented:
Travis error is unrelated.
